### PR TITLE
[js-api] Add stack trace support and preserve identity of JS exception values passing through WebAssembly.

### LINKED ISSUE #189: Preserve identity of exceptions from JS when passing through wasm
Some discussion starting at https://github.com/WebAssembly/exception-handling/issues/183#issuecomment-955090191 already.

Proposal: attach an optional `externref` to the core exception, and make sure it's propagated by `catch`; `rethrow`, but not expose it to wasm code. Then hosts that don't need it, should be able to avoid storing it at all. The JS API spec can then extract that value when it gets an exception from wasm, or create a new one if there isn't one.

### LINKED ISSUE #201: Port stack trace support from explainer to js api spec
#197

### PR DESCRIPTION
**Pull request by Ms2ger**
Fixes #189.
Fixes #201.

### CODE REVIEW

File: document/js-api/index.bs [hunk 1]
```diff
@@ -1043,11 +1043,11 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
     1. Note: The expectation is that [=func_invoke=] will be updated to return (|store|, <var ignore>val</var>* | [=error=] | (exception |exntag| |payload|)).
```
Comment (aheejin) [inline document/js-api/index.bs]: Shouldn't this line have `opaqueData` too?

File: document/js-api/index.bs [hunk 2]
```diff
@@ -1105,10 +1105,16 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
             1. If |v| [=implements=] {{Exception}},
                 1. Let |type| be |v|.\[[Type]].
                 1. Let |payload| be |v|.\[[Payload]].
+                1. If |v|.\[[Stack]] is undefined,
+                    1. Let |opaqueValue| be [=ref.null=] [=externref=].
```
Comment (aheejin) [inline document/js-api/index.bs]: We use `opaqueData` in another paragraph and `opaqueValue` here. Are they different entities?

File: document/js-api/index.bs [hunk 3]
```diff
@@ -1043,11 +1043,11 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
     1. Note: The expectation is that [=func_invoke=] will be updated to return (|store|, <var ignore>val</var>* | [=error=] | (exception |exntag| |payload|)).
     1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
     1. If |ret| is [=error=], throw an exception. This exception should be a WebAssembly {{RuntimeError}} exception, unless otherwise indicated by <a href="#errors">the WebAssembly error mapping</a>.
-    1. If |ret| is exception |exntag| |payload|, then
+    1. If |ret| is exception |exntag| |payload| |opaqueData|, then
         1. If |exntag| is the [=JavaScript exception tag=], then
             1. Let « [=ref.extern=] |externaddr| » be |payload|.
             1. Throw the result of [=retrieving an extern value=] from |externaddr|.
```
Comment (aheejin) [inline document/js-api/index.bs]: Do we not need to throw `opaqueData` together here?

File: document/js-api/index.bs [hunk 4]
```diff
@@ -1105,10 +1105,16 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
+                1. If |v|.\[[Stack]] is undefined,
+                    1. Let |opaqueValue| be [=ref.null=] [=externref=].
+                1. Otherwise,
+                    1. Assert: [=Type=](|v|.\[[Stack]]) is String.
+                    1. Let |opaqueValue| be [=ToWebAssemblyValue=](|v|.\[[Stack]], [=externref=])
             1. Otherwise,
                 1. Let |type| be the [=JavaScript exception tag=].
                 1. Let |payload| be [=ToWebAssemblyValue=](|v|, [=externref=]).
-            1. [=WebAssembly/Throw=] with |type| and |payload|.
+                1. Let |opaqueValue| be [=ref.null=] [=externref=].
```
Comment (aheejin) [inline document/js-api/index.bs]: Does this include when something that's not a `WebAssembly.Exception`, let's say, `RuntimeError` or just a plane JS exception, such as JS's `Error` or a plain value (e.g. `throw 3`)? In these cases, is it correct to consider them to have a "Javascript exception tag", as if it is an instance of `WebAssembly.Exception`?

File: document/js-api/index.bs [hunk 5]
```diff
@@ -1258,20 +1264,25 @@ Advisement: This method is only expected to be implemented or shipped when both
...
 
 <pre class="idl">
+dictionary ExceptionOptions {
+  boolean traceStack = false;
+};
+
 [LegacyNamespace=WebAssembly, Exposed=(Window,Worker,Worklet)]
 interface Exception {
-  constructor(Tag exceptionTag, sequence&lt;any> payload);
+  constructor(Tag exceptionTag, sequence&lt;any> payload, optional ExceptionOptions options = {});
```
Comment (aheejin) [inline document/js-api/index.bs:1268]: Does an optional argument need a default parameter (`{}`)?

File: document/js-api/index.bs [hunk 6]
```diff
@@ -1258,20 +1264,25 @@ Advisement: This method is only expected to be implemented or shipped when both
...
 </pre>
 
 An {{Exception}} value represents an exception.
 
 <div algorithm>
 
-To <dfn>create an Exception object</dfn> from a [=tag address=] |tagAddress| and [=list=]
-of WebAssembly values |payload|, perform the following steps:
+To <dfn>create an Exception object</dfn> from a [=tag address=] |tagAddress|, [=list=] of
+WebAssembly values |payload|, and [=externref=] |opaqueData|, perform the following steps:
```
Comment (aheejin) [inline document/js-api/index.bs]: Does this happen on the JS side? What's the difference between this algorithm and "The `new Exception(exceptionTag, payload, options)` constructor steps are:" below?

**Review by aheejin — COMMENTED** [review]
Thanks! Sorry for the late reply; I was OOO.

Sorry for the many questions; I am not very familiar with the JS API text.

Other questions:
1. We have sections called "Call an exported function" and "Create a host function". Does "Call an exported function" happen on the JS side and does "Create a host function" on the Wasm side?

2. Another question about the existing issue:
```
Issue: Should it be possible for `catch <JS-exception-tag>` to extract the payload from an exception with this tag?
```
Can we consider this resolved given that we have an JS API that can create `WebAssembly.Exception`?

3. Does this also address #189, or does it need to be separately addressed?

**Comment by Ms2ger** [comment]
1. "Call an exported function" is calling a wasm function from js; "Create a host function" is calling a js function from wasm.
2. Consider the JS function `f` defined as `function f() { throw {} }` - that is, `f` throws an exception value that is **not** a `WebAssembly.Exception`. The question is, if `f` is called from webassembly, whether the webassembly code should be able to extract an `externref` from that exception. This is not yet supported, but I'm thinking that it could be deferred to a future version of wasm.
3. #189 is not solved yet.

However, I'm realizing I don't fully understand how `traceStack` is supposed to work. The explainer has a line "`traceStack` serves as a request to the WebAssembly engine to attach a stack trace", but I don't know what that's supposed to look like. We pass the option to the JS `Exception` constructor and capture a stack at that point, but that doesn't change the behavior of the wasm engine. Or is the expectation that in JS code called from wasm, an `Exception` object will have a stack trace that either contains only JS frames or interleaved JS and wasm frames?

File: document/js-api/index.bs (hunk 5 above)
Comment (Ms2ger) [reply document/js-api/index.bs:1268]: Yes.

File: document/js-api/index.bs (hunk 1 above)
Comment (Ms2ger) [reply document/js-api/index.bs]: Good catch, fixed.

File: document/js-api/index.bs (hunk 2 above)
Comment (Ms2ger) [reply document/js-api/index.bs]: It's just a different variable name, but I made them consistent.

File: document/js-api/index.bs (hunk 6 above)
Comment (Ms2ger) [reply document/js-api/index.bs]: This is the algorithm that handles exceptions thrown from inside WebAssembly itself, as opposed to one created in JS.

**Comment by Ms2ger** [comment]
Thanks for making me think through #189; this PR now fixes that as well.

File: document/js-api/index.bs [hunk 7]
```diff
@@ -1107,8 +1107,9 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
                 1. Let |payload| be |v|.\[[Payload]].
             1. Otherwise,
                 1. Let |type| be the [=JavaScript exception tag=].
-                1. Let |payload| be [=ToWebAssemblyValue=](|v|, [=externref=]).
-            1. [=WebAssembly/Throw=] with |type| and |payload|.
+                1. Let |payload| be « ».
+            1. Let |opaqueData| be [=ToWebAssemblyValue=](|v|, [=externref=])
```
Comment (aheejin) [inline document/js-api/index.bs:1111]: It looks this wraps the whole value itself as `opaqueData`, correct?  Where does the stack trace belong then?

File: document/js-api/index.bs [hunk 8]
```diff
@@ -1107,8 +1107,9 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
                 1. Let |payload| be |v|.\[[Payload]].
             1. Otherwise,
                 1. Let |type| be the [=JavaScript exception tag=].
-                1. Let |payload| be [=ToWebAssemblyValue=](|v|, [=externref=]).
-            1. [=WebAssembly/Throw=] with |type| and |payload|.
+                1. Let |payload| be « ».
```
Comment (aheejin) [inline document/js-api/index.bs:1110]: Do we not need to handle `|v|.\[[Stack]]` in here and also in line 1106-1107?

File: document/js-api/index.bs [hunk 9]
```diff
@@ -1040,12 +1040,12 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
...
     1. Let (|store|, |ret|) be the result of [=func_invoke=](|store|, |funcaddr|, |args|).
-    1. Note: The expectation is that [=func_invoke=] will be updated to return (|store|, <var ignore>val</var>* | [=error=] | (exception |exntag| |payload|)).
+    1. Note: The expectation is that [=func_invoke=] will be updated to return (|store|, <var ignore>val</var>* | [=error=] | (exception |exntag| |payload| |opaqueData|)).
     1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
     1. If |ret| is [=error=], throw an exception. This exception should be a WebAssembly {{RuntimeError}} exception, unless otherwise indicated by <a href="#errors">the WebAssembly error mapping</a>.
-    1. If |ret| is exception |exntag| |payload|, then
-        1. If |exntag| is the [=JavaScript exception tag=], then
-            1. Let « [=ref.extern=] |externaddr| » be |payload|.
+    1. If |ret| is exception |exntag| |payload| |opaqueData|, then
+        1. If |opaqueData| is not [=ref.null=] [=externref=],
```
Comment (aheejin) [inline document/js-api/index.bs:1047]: In which case is this `ref.null`? When it is thrown from wasm?

**Review by aheejin — COMMENTED** [review]
1. Create a new exported function
2. Call an exported function
3. Create a host function
4. Run a host function

We only handle exceptions in 2 and 3. I think I don't have a clear concept of the difference between "create" and "call/run"; the steps look kind of similar to me. Why do we only need to handle "call" in case of an exported function and "create" in case of a host function?

File: document/js-api/index.bs (hunk 9 above)
Comment (Ms2ger) [reply document/js-api/index.bs:1047]: Indeed.

File: document/js-api/index.bs (hunk 7 above)
Comment (Ms2ger) [reply document/js-api/index.bs:1111]: The stack trace just stays in a slot on the `Exception` JS object

**Review by Ms2ger — COMMENTED** [review]
The split between call/run/create is not particularly meaningful. In the "host function" case, it exists because there's no well-defined try-finally construct in prose, so "run a host function" exists as an algorithm you can early-return from if there's an exception, and then "create a host function" can put the (exception-or-normal) result in a variable. I think we could merge "call an Exported Function" into "a new Exported function" now that ECMAScript uses abstract closures for CreateBuiltinFunction (https://github.com/tc39/ecma262/issues/1894); the current split predates that (https://github.com/WebAssembly/spec/commit/d15cabb244420016058d826cbcf8e469c14782fe).

**Comment by aheejin** [comment]
@dschuff Do you have any comments?

**Comment by dschuff** [comment]
Thinking about the exception identity here... IIUC:
When an exception is originated from wasm:
1. the first time it propagates into JS `opaqueData` is null (so it lands in 12.2 of "call an Exported Function") and its tag and payload get wrapped in a new WebAssembly.Exception and thrown.
3. If it propagates back into wasm, it will land in 3.9.2 of "create a host function". The type and payload are taken out of the object and a reference to the object itself goes into `opaqueData`.
4. If it then propagates back into JS again, it will land in 12.1, and `opaqueData` (the original wrapping from 1) gets thrown.


When an exception is originated from JS:
1. the first time it propagates into wasm, it will land in 3.9.3 of "create a host function",  the type will be the JS tag, the payload will be empty, and `opaqueData` will be the original thrown object. 
2. If it propagates back into JS, it will be like 3 above (landing in 12.1 of "call an exported function")

One thing I can't recall from our discussion: when exceptions propagate into wasm (i.e. where we say "To throw" in section 4.7.3) it says "invoke the catch block with payload and opaqueData". What does that mean? Which of the payload and/or opaqueData gets actually pushed onto the wasm value stack in the catch block?

**Comment by Ms2ger** [comment]
> Thinking about the exception identity here... IIUC:

All this matches my understanding.

> One thing I can't recall from our discussion: when exceptions propagate into wasm (i.e. where we say "To throw" in section 4.7.3) it says "invoke the catch block with payload and opaqueData". What does that mean? Which of the payload and/or opaqueData gets actually pushed onto the wasm value stack in the catch block?

I believe the goal is:

* the opaque data is not exposed to wasm code (in fact, if your wasm implementation doesn't interoperate with JS, the opaque data can be ignored entirely)
* `catch` (assuming the tag matches) pushes the payload onto the stack and keeps track of the opaque data separately
* `rethrow` grabs the opaque data from wherever `catch` put it, so that `catch`+`rethrow` is a no-op

I think it would be best if this was clarified in the core spec, but I haven't had time to dig into where exactly that would go.

**Comment by aheejin** [comment]
The current formal reduction rule for `rethrow` in the core spec is:
```
caught{a val*} B^l[rethrow l] end
  ↪ caught{a val*} B^l[val* (throw a)] end
```
So this does a fresh throw, which I think can't be helped in the formal spec because formal spec doesn't have a concept of opaque data. The core spec only knows about the tag and the payload, so there's no other way to express the functionality anyway.

But should we say something about the core spec's `rethrow` here in the JS API, because apparently a fresh throw within the wasm is not what we want? This might sound weird, but not sure if there's any other way around this. cc @ioannad

**Comment by aheejin** [comment]
@Ms2ger Gentle ping for my latest question :)

**Comment by ioannad** [comment]
> The current formal reduction rule for `rethrow` in the core spec is:
> 
> ```
> caught{a val*} B^l[rethrow l] end
>   ↪ caught{a val*} B^l[val* (throw a)] end
> ```
> 
> So this does a fresh throw, which I think can't be helped in the formal spec because formal spec doesn't have a concept of opaque data. The core spec only knows about the tag and the payload, so there's no other way to express the functionality anyway.
> 
> But should we say something about the core spec's `rethrow` here in the JS API, because apparently a fresh throw within the wasm is not what we want? This might sound weird, but not sure if there's any other way around this. cc @ioannad

Indeed, according to the current formal spec, `caught` does only store the tag and payload. However, I consider `throw tagidx` a fresh throw, not the administrative `throw tagaddr`, which could be seen as just initiating the throwing process internally. 

I would have to think of a different way to address opaque data in the core spec, if that is even necessary.
Any thoughts or ideas on this,  @rossberg ?

**Comment by Ms2ger** [comment]
Apologies for the delay - I haven't had much wasm time recently. I'd prefer landing this PR as is, and I'll work with @ioannad to update the core spec to clarify how the opaque data is passed through.

**Review by aheejin — APPROVED** [review]

**Comment by rossberg** [comment]
The core spec doesn't have any notion of exception identity – we removed that intentionally, otherwise we could have kept `exnref`. To model such an identity, you'll basically have to introduce exnref internally, but I'm not sure how desirable it is to reintroduce it through the backdoor.

So with the current state of affairs, I think it rather ought to be hand-waved in the JS API spec somehow.

(Of course, I'd be all for reintroducing exnref properly, which would allow to significantly simplify the proposal and solve the known expressiveness problems. :) )

**Comment by Ms2ger** [comment]
I'm going to merge this as is, as it's a significant improvement over the status quo, and I'll file an issue on either making the changes to the core spec or hand-waving in the js-api spec.

**Pull request merged**